source/common/upstream/cluster_manager_impl.cc

```diff
@@ -877,11 +879,12 @@ void ClusterManagerImpl::clusterWarmingToActive(const std::string& cluster_name)
   warming_clusters_.erase(warming_it);
 }
 
-bool ClusterManagerImpl::removeCluster(const std::string& cluster_name) {
+bool ClusterManagerImpl::removeCluster(const std::string& cluster_name, const bool remove_ignored) {
   bool removed = false;
   auto existing_active_cluster = active_clusters_.find(cluster_name);
   if (existing_active_cluster != active_clusters_.end() &&
-      existing_active_cluster->second->added_via_api_) {
+      existing_active_cluster->second->added_via_api_ &&
```

Comment:
Thanks for picking up this change, and sorry coming in late with a possibly naive question.
So the problem is we add a DFP cluster via CDS, the DFP cluster adds each host as a sub-cluster, and subsequently the DFP sub-clusters are removed by CDS updates right?

Would it be cleaner rather than adding the ignore tags, to simply have the sub-clusters not tagged as added via API?  Or would that then cause problems for removal?

If it complicates removal WDYT of instead of having an "added by API" bool and a "ignore" bool, if we change both addOrUpdateCluster to include a enum Type (Static, Cds, DFP) where each cluster retains its type and we ignore CDS removals for DFP?

/wait-any for comments

Reply:
Thanks for your inputs. 

Tagging clusters with `not added_via_api` would make the changes more complex and could impact existing behavior. 

The cluster api spec already has `type` and `cluster_type` fields, but the DFP sub-cluster sets `type` to `STRICT_DNS`. Having another enum in API spec or having enum as addOrUpdateCluster param will abstract the resulting behavior, i.e., avoiding cds removal.

I feel `avoid_cds_removal` parameter makes it clear & clean for the caller and easy to manage, not just for DFP sub-clusters, but for any other use cases in future where custom implementations will manage the cluster lifecycle (similar to DFP).

Open for thoughts :)